### DISCUSSION TITLE
Use Avalonia for commercial project

### DISCUSSION THREAD

**Discussion by OmidID** [AvaloniaUI/Avalonia#11799] [General]
We want to start a new project that will be work on LINUX + Custom hardware. Do you think is it good idea to use Avalonia for GUI? 

Maybe we will have some complicated GUI.

**Comment by danwalmsley** [comment]
there are a few commercial applications in development using Avalonia. Things are fairly stable now.

You may encounter the odd issue, so be prepared for that. It also depends on the processing power of your custom hardware.

**Comment by OmidID** [comment]
@danwalmsley , Thanks,
This is the CPU we using:
https://www.toradex.com/computer-on-modules/colibri-arm-family/nvidia-tegra-2

**Comment by legistek** [comment]
I didn't want to open a new/issue question for this, since it's so closely related, so -

Are there any commercial applications built on Avalonia that you can cite for those of us looking for a new UI framework who might have to answer to risk-averse investors, etc.?

**Reply by kekekeks** [reply]
See https://avaloniaui.net/Showcase

**Reply by robloo** [reply]
@legistek 

Icons8 Lunacy, Unity Plastic SCM and GritWorld GritGene are the major apps on the showcase list. These are about as big and complex as apps get and have been using Avalonia for years. These are top tier. Throw a few of these examples in the investor deck and it will he hard to argue against.

Other apps are great too but are not at this scale from what I've seen.

**Reply by OmidID** [reply]
Very old conversation :) 
In fact, we reach production using iMX6 at that time and we are much more productive now compare to 5 years ago. 
I do confirm we used Avalonia in many products. 
Good luck

**Reply by danipen** [reply]
We at Unity Version Control (formerly Plastic SCM) are using Avalonia since version 0.10.4. The desktop client is quite big. All went smooth since the beginning . We just support desktop (win/osx/Linux).

**Reply by llfab** [reply]
At Stryker (www.stryker.com) we use Avalonia for intra operative medical applications currently on Windows only but exactly for the reason of being platform independent.

**Reply by legistek** [reply]
Thank you all for the info!

**Comment by mysteryx93** [comment]
Custom hardware should be a Linux concern, not an Avalonia concern.

You can run some of the showcase Avalonia applications and see if you get any issue at all.

Also for a commercial project with investors, you might want to consider getting commercial support from Avalonia team. If there's any hardware-specific bug you encounter, the team will work with you to resolve those corner-case bugs.

If you do that, I don't see any risk.

**Reply by llfab** [reply]
I can only encourage paid support for two reasons:
- you will get your private chat channel to the team to solve whatever problem quickly
- only if more enterprise users give money to the Avalonia core team we all make sure that the team can grow bringing the system forward

**Reply by legistek** [reply]
Well since the alternatives I'm considering include a custom UI built off of DirectComposition, support isn't really a concern. :) But it does help to be able to say "so and so" use it too. What I need is something that's: (a) .NET-based; (b) MVVM-friendly; (c) uses DirectX; and (d) not built off of WinRT mickey mouse garbage. That limits the options severely.

**Reply by llfab** [reply]
- .NET based: check
- MVVM-friendly: check
- Not WinRT: check
- DX: AvaloniaUI rendering is based on Skia (like Firefox etc.) and by default renders with OpenGL. There is a Vulkan backend too for Skia but its fairly new. If you need to integrate your own DX renderings that should be easily possible by Avalonia's External GPU Memory integration (we integrate our own Vulkan 3D renderer using that API. So, were comes your need for DX from?

**Reply by legistek** [reply]
Well you have the D2D renderer don't you? 
I have a proprietary video player built off of ffmpeg/libav (similar in concept to VLC) that I use with WPF that outputs to a DXGI surface. As long as I can get it onto your swap chain without leaving VRAM I'm good. I'm also not at all above forking and extending the Direct2D renderer if I need to since you've already done most of the hard work.

**Reply by llfab** [reply]
DXGI integration into Avalonia on Skia on OpenGL is not a problem using external GPU memory API. We are doing the same thing but with Vulkan textures not leaving VRAM

**Reply by llfab** [reply]
Not 100% sure about the D2D renderers state but I believe default composition renderer and the mentioned texture sharing would potentially be the better setup

**Reply by legistek** [reply]
Awesome, good to know thank you!

**Reply by mysteryx93** [reply]
DirectX on Linux? Good luck with that!

There's no DirectX in Linux :) other than DXVK

**Reply by legistek** [reply]
Sorry I was only asking about Windows.

**Reply by OmidID** [reply]
Why not use Linux? So you can use libdrm and basically boot your OS to your application directly. I really don't know what is your project and what you want to do, but i expect, you want to have some embedded devices. Correct me if i am wrong.

**Reply by legistek** [reply]
My commercial segment is business users on Windows desktops. :)

So you can imagine my frustration/indignation at how badly Microsoft has botched that developer ecosystem!

**Reply by OmidID** [reply]
OK. if you want to run desktop app you do not need to run it over DX. you simply can use Skia and OpenGL and everything works perfectly. i have some apps working perfectly on windows.

**Reply by legistek** [reply]
Is there something wrong with the D2D renderer that I'm missing? I'm just wondering why everyone is trying to steer me away from it.

**Reply by jp2masa** [reply]
Skia is cross-platform and works well, so D2D renderer development is not very active. IIRC, it was considered to be removed at some point, but some people are using it and it works fine, so it probably won't be removed soon. I may be wrong, though. So, the recommended approach is to use Skia unless you really need D2D for some reason.

**Reply by OmidID** [reply]
You are correct @jp2masa .
@legistek  You can read the reason here https://github.com/sharpdx/SharpDX

**Reply by legistek** [reply]
Oh because SharpDX was canned? Gotcha. 
Well I have zero problem forking the D2D renderer and updating it if need be, if that turns out to be the best approach.
Thanks all.

**Reply by llfab** [reply]
I believe that many of the new features (e.g., render thread animations) in Avalonia are based on the composition renderer that uses Skia. So, overall you may be better off going the default path. There should be good enough samples for the DXGI interop.

**Reply by legistek** [reply]
@llfab you're talking about, among others, this I assume?

https://github.com/AvaloniaUI/Avalonia/tree/1b9aba1cf0f910edb9d463a6b757892c4a732fff/samples/GpuInterop

**Reply by llfab** [reply]
Yes, we are using that API for Vk 3D renderer integration

**Reply by legistek** [reply]
I got my video player working (quite well in fact) using the rough pattern in the GpuInterop example. Thanks all!

**Reply by llfab** [reply]
> I got my video player working (quite well in fact) using the rough pattern in the GpuInterop example. Thanks all!

Good job! Not the easiest bit of code I guess
